modules/katamari/src/main/ts/ephox/katamari/api/Optional.ts

```diff
@@ -1,141 +1,290 @@
 
-  /** Same as "or", but uses a thunk instead of a value */
-  readonly orThunk: <T2 = T>(makeOption: () => Optional<T2>) => Optional<T | T2>;
+  /**
+   * Determine if this `Optional` object **does not** contain a value.
+   */
+  public isNone(): boolean {
+    return !this.tag;
+  }
 
-  /** Run a function over the 'some' value.
-   *  "map" operation on the Optional functor.
+  // --- Functor (name stolen from Haskell / maths) ---
+
+  /**
+   * Perform a transform on an `Optional` object, **if** there is a value. If
+   * you provide a function to turn a T into a U, this is the function you use
+   * to turn an `Optional<T>` into an `Optional<U>`. If this **does** contain
+   * a value then the output will also contain a value (that value being the
+   * output of `mapper(this.value)`), and if this **does not** contain a value
+   * then neither will the output.
    */
-  readonly map: <U> (mapper: (x: T) => U) => Optional<U>;
+  public map<U>(mapper: (value: T) => U): Optional<U> {
+    if (this.tag) {
+      return Optional.some(mapper(this.value as T));
+    } else {
+      return Optional.none();
+    }
+  }
 
-  /** Run a side effect over the 'some' value */
-  readonly each: (worker: (x: T) => void) => void;
+  // --- Monad (name stolen from Haskell / maths) ---
 
-  /** "bind"/"flatMap" operation on the Optional Bind/Monad.
-   *  Equivalent to >>= in Haskell/PureScript; flatMap in Scala.
+  /**
+   * Perform a transform on an `Optional` object, **if** there is a value.
+   * Unlike the map function earlier in the piece, here the transform itself
+   * also returns an `Optional`.
    */
-  readonly bind: <U> (f: (x: T) => Optional<U>) => Optional<U>;
+  public bind<U>(binder: (value: T) => Optional<U>): Optional<U> {
+    if (this.tag) {
+      return binder(this.value as T);
+    } else {
+      return Optional.none();
+    }
+  }
 
-  /** Does this Optional contain a value that predicate? */
-  readonly exists: (f: (x: T) => boolean) => boolean;
+  // --- Traversable (name stolen from Haskell / maths) ---
 
-  /** Do all values contained in this option match this predicate? */
-  readonly forall: (f: (x: T) => boolean) => boolean;
+  /**
+   * For a given predicate, this function finds out if there **exists** a value
+   * inside this `Optional` object that meets the predicate. In practice, this
+   * means that for empty `Optional`s it returns false (as no predicate-
+   * meeting value exists).
+   */
+  public exists(predicate: (value: T) => boolean): boolean {
+    return this.tag && predicate(this.value as T);
+  }
 
-  /** Return all values in this Optional that match the predicate.
-   *  The predicate may refine the constituent type using TypeScript type predicates.
+  /**
+   * For a given predicate, this function finds out if **all** the values inside
+   * this `Optional` object meet the predicate. In practice, this means that
+   * for empty `Optional`s it returns true (as all 0 objects do meet the
+   * predicate).
```

Comment:
Similar to my previous comment, using empty is not clear.
```suggestion
  /**
   * For a given predicate, this function finds out if **all** the values inside
   * this `Optional` object meet the predicate. In practice, this means that
   * for a `None` it returns true (as all 0 objects do meet the predicate).
```

Reply:
As above